aggregator/src/aggregator/key_rotator.rs

```diff
@@ -0,0 +1,673 @@
+                tx.set_global_hpke_keypair_state(keypair.config().id(), &HpkeKeyState::Active)
+                    .await?;
+            }
+        } else {
+            let to_be_expired_keypairs: Vec<_> = active_keypairs
+                .iter()
+                .filter(|keypair| {
+                    Self::duration_since(tx, keypair.updated_at()) > config.active_duration
+                })
+                .cloned()
+                .collect();
+
+            if to_be_expired_keypairs != active_keypairs {
+                for keypair in to_be_expired_keypairs {
+                    info!(id = ?keypair.id(), "allowing already active keypair to replace this one");
+                    tx.set_global_hpke_keypair_state(keypair.id(), &HpkeKeyState::Expired)
+                        .await?;
+                }
+            } else {
+                if pending_keypairs.is_empty() {
+                    let next = next_key()?;
+                    info!(id = ?next.config().id(), "inserting new pending key to replace expired one(s)");
+                    tx.put_global_hpke_keypair(&next).await?;
+                } else {
+                    let pending_key_is_ready = pending_keypairs.iter().any(|keypair| {
+                        Self::duration_since(tx, keypair.updated_at()) > config.pending_duration
+                    });
+                    if pending_key_is_ready {
+                        for keypair in to_be_expired_keypairs {
+                            info!(id = ?keypair.id(), "a pending key is ready, marking key expired");
+                            tx.set_global_hpke_keypair_state(keypair.id(), &HpkeKeyState::Expired)
+                                .await?;
+                        }
+                    }
+                }
+            }
+
+            // Promote any pending keypairs that are ready. We don't sweep them in the same loop
+            // as active keypairs, because it's possible for there to be a pending keypair
+            // without an active one.
+            for pending_keypair in pending_keypairs {
+                if Self::duration_since(tx, pending_keypair.updated_at()) > config.pending_duration
+                {
+                    info!(id = ?pending_keypair.id(), "pending key is ready, moving it to active");
+                    tx.set_global_hpke_keypair_state(pending_keypair.id(), &HpkeKeyState::Active)
+                        .await?;
+                }
+            }
+        }
+
+        // Sweep expired keypairs. Any keypairs that are old enough are deleted.
+        for keypair in &expired_keypairs {
+            if Self::duration_since(tx, keypair.updated_at()) > config.expired_duration {
+                info!(id = ?keypair.id(), "deleting expired keypair");
+                tx.delete_global_hpke_keypair(keypair.id()).await?;
+            }
+        }
+
+        Ok(())
+    }
+
+    fn duration_since(tx: &Transaction<C>, time: &Time) -> Duration {
```

Comment:
nit: I think making this take a `&C` argument instead would be a bit clearer, as it would be obvious there's no database manipulation happening. You could also add a `let clock = tx.clock();` binding at the top of `run_hpke_for_config()` to eliminate some repetition.